dot/sync/chain_processor.go

```diff
@@ -95,99 +95,109 @@ func (s *chainProcessor) processReadyBlocks() {
 // processBlockData processes the BlockData from a BlockResponse and
 // returns the index of the last BlockData it handled on success,
 // or the index of the block data that errored on failure.
-func (s *chainProcessor) processBlockData(bd *types.BlockData) error {
-	if bd == nil {
-		return ErrNilBlockData
-	}
+func (c *chainProcessor) processBlockData(blockData types.BlockData) error { //nolint:revive
+	logger.Debugf("processing block data with hash %s", blockData.Hash)
 
-	hasHeader, err := s.blockState.HasHeader(bd.Hash)
+	headerInState, err := c.blockState.HasHeader(blockData.Hash)
 	if err != nil {
-		return fmt.Errorf("failed to check if block state has header for hash %s: %w", bd.Hash, err)
+		return fmt.Errorf("checking if block state has header: %w", err)
```

Comment:
Technically you already checked.

```suggestion
		return fmt.Errorf("Block state has no header: %w", err)
```

Reply:
- the error comes from `c.blockState.HasHeader` and, if non-nil, does not mean the block state has no header for that block hash. It just failed checking if the block state had the header (i.e. disk io error etc.).
- the error occurred whilst `checking if block state has header`. The bottom wrapped error usually describes what actually went wrong. The `checking if block state has header: ` is just context (or human readable stack trace if you prefer) for that wrapped error.
- Don't capitalize errors :wink: see https://github.com/golang/go/wiki/CodeReviewComments#error-strings